### ISSUE TITLE
End-to-End Training Validation: From Raw Data to Verified Model

### ISSUE THREAD

**Issue by akaszubski** [akaszubski/realign#51]
## Summary

Implement end-to-end validation workflow that tests the complete ReAlign pipeline from raw data curation through training to verified, benchmarked model. Validates that each content type training produces measurable improvements.

### Why End-to-End Validation?

**Quality Assurance**:
- Prove ReAlign pipeline works as designed
- Validate training method recommendations from research
- Catch regressions early
- Demonstrate value to users

**Evidence-Based**:
- Show measurable improvement for each content type
- Validate multi-stage training approach
- Prove distributed training benefits
- Document best practices

**User Confidence**:
- Working examples for all 10 content types
- Reproducible results
- Clear success metrics
- Troubleshooting guidance

### Phase 1: Test Dataset Creation

Create small, focused test datasets for each content type (100-500 examples each).

- [ ] **Create test datasets** (`data/test_datasets/`)
  ```
  data/test_datasets/
  ├── belief_test/
  │   ├── train.jsonl (Constitutional AI examples)
  │   └── eval.jsonl (Ethical dilemmas)
  ├── fact_test/
  │   ├── train.jsonl (Medical facts)
  │   └── eval.jsonl (Fact recall questions)
  ├── reasoning_test/
  │   ├── train.jsonl (Math problems with solutions)
  │   └── eval.jsonl (New math problems)
  └── ... (all 10 types)
  ```

- [ ] **Document dataset creation**
  - Sources for each test set
  - Format specifications
  - Quality criteria
  - Expected difficulty

### Phase 2: Baseline Measurements

Establish baseline performance before training for each content type.

- [ ] **Create baseline runner** (`src/realign/validation/baseline_runner.py`)
  ```python
  class BaselineRunner:
      def run_baseline_tests(self, model: str, content_type: str) -> Dict:
          """Run baseline tests for content type before training"""
          
      def save_baseline(self, results: Dict, output_path: str):
          """Save baseline results for comparison"""
  ```

- [ ] **Run baselines for all content types**
  ```bash
  python -m realign.validation.baseline \
    --model meta-llama/Llama-3.1-70B \
    --content-types all \
    --output baselines/llama-70b.json
  ```

- [ ] **Baseline metrics per content type**
  - BELIEF: Constitutional AI alignment score
  - FACT: Factual accuracy (F1, exact match)
  - PROCEDURE: Step completeness, ordering accuracy
  - SKILL: Quality score (UltraFeedback-style)
  - REASONING: Problem-solving accuracy (GSM8K-style)
  - BEHAVIOR: Consistency score across contexts
  - PREFERENCE: Style adherence score
  - CONSTRAINT: Safety refusal rate
  - CONTEXT: Domain knowledge accuracy
  - EXPERIENCE: Analogical reasoning accuracy

### Phase 3: Training Execution

Execute training for each content type using recommended method.

- [ ] **Create training executor** (`src/realign/validation/training_executor.py`)
  ```python
  class TrainingExecutor:
      def train_content_type(
          self, 
          model: str, 
          content_type: str,
          dataset: str,
          method: str  # From TRAINING_METHODS_BY_CONTENT_TYPE.md
      ) -> str:
          """Train model on content type, return adapter path"""
          
      def train_multi_stage(
          self,
          model: str,
          training_plan: Dict  # From #49 content type classifier
      ) -> str:
          """Execute multi-stage training plan"""
  ```

- [ ] **Training configurations per content type**
  ```yaml
  # configs/validation/belief_training.yaml
  content_type: BELIEF
  method: dpo
  dataset: data/test_datasets/belief_test/train.jsonl
  config:
    beta: 0.1
    learning_rate: 5e-6
    iterations: 500
    batch_size: 4
  expected_improvement: 0.15  # 15% improvement over baseline
  ```

- [ ] **Execute all trainings**
  ```bash
  # Single content type
  python -m realign.validation.train \
    --content-type BELIEF \
    --config configs/validation/belief_training.yaml \
    --output models/validated/belief

  # All content types
  python -m realign.validation.train_all \
    --base-model meta-llama/Llama-3.1-70B \
    --output-dir models/validated/
  ```

### Phase 4: Post-Training Verification

Measure improvement after training for each content type.

- [ ] **Create post-training evaluator** (`src/realign/validation/post_training_evaluator.py`)
  ```python
  class PostTrainingEvaluator:
      def evaluate_improvement(
          self,
          trained_model: str,
          content_type: str,
          baseline_results: Dict
      ) -> Dict:
          """Compare trained model to baseline"""
          
      def detect_regressions(
          self,
          trained_results: Dict,
          baseline_results: Dict
      ) -> List[str]:
          """Detect performance regressions"""
          
      def calculate_improvement(
          self,
          trained_results: Dict,
          baseline_results: Dict
      ) -> float:
          """Calculate percentage improvement"""
  ```

- [ ] **Run post-training evaluations**
  ```bash
  python -m realign.validation.evaluate \
    --trained models/validated/belief \
    --baseline baselines/llama-70b.json \
    --content-type BELIEF \
    --output results/validated/belief.json
  ```

- [ ] **Success criteria per content type**
  ```python
  SUCCESS_CRITERIA = {
      "BELIEF": {
          "min_improvement": 0.15,  # 15% better alignment
          "max_regression": 0.05,   # Max 5% regression elsewhere
          "required_metrics": ["alignment_score", "consistency"]
      },
      "FACT": {
          "min_improvement": 0.20,  # 20% better accuracy
          "max_regression": 0.03,
          "required_metrics": ["accuracy", "hallucination_rate"]
      },
      # ... for all 10 types
  }
  ```

### Phase 5: Regression Detection

Ensure training on one content type doesn't hurt others (catastrophic forgetting).

- [ ] **Implement regression detector** (`src/realign/validation/regression_detector.py`)
  ```python
  class RegressionDetector:
      def detect_catastrophic_forgetting(
          self,
          trained_model: str,
          content_type_trained: str,
          baseline_results: Dict
      ) -> Dict:
          """Test if training hurt other content types"""
          
      def run_cross_content_tests(
          self,
          trained_model: str
      ) -> Dict:
          """Test all content types after training one"""
  ```

- [ ] **Cross-content type testing**
  ```bash
  # After training on BELIEF, test all other types
  python -m realign.validation.regression \
    --trained models/validated/belief \
    --baseline baselines/llama-70b.json \
    --test-all-types \
    --output results/regression/belief_training.json
  ```

- [ ] **Regression report**
  - Which content types regressed?
  - By how much?
  - Is it within acceptable limits?
  - Recommendations for mitigation

### Phase 6: Multi-Stage Training Validation

Validate that multi-stage training (combining content types) works better than single-stage.

- [ ] **Implement multi-stage validator** (`src/realign/validation/multistage_validator.py`)
  ```python
  class MultiStageValidator:
      def validate_training_plan(self, plan: Dict) -> Dict:
          """Execute and validate multi-stage training plan"""
          # Stage 1: Facts + Context (LoRA)
          # Stage 2: Skills + Experience (LoRA→DPO)
          # Stage 3: Beliefs + Behaviors (DPO)
          # Stage 4: Constraints (DPO high beta)
          # Stage 5: Reasoning (GRPO)
          
      def compare_strategies(
          self,
          single_stage_results: Dict,
          multi_stage_results: Dict
      ) -> Dict:
          """Compare single-stage vs multi-stage training"""
  ```

- [ ] **Test training strategies**
  ```bash
  # Single-stage (baseline)
  python -m realign.validation.train_single_stage \
    --model llama-70b \
    --all-content-types \
    --method lora \
    --output models/single_stage

  # Multi-stage (recommended)
  python -m realign.validation.train_multi_stage \
    --model llama-70b \
    --plan configs/validation/multistage_plan.yaml \
    --output models/multi_stage

  # Compare
  python -m realign.validation.compare \
    --model1 models/single_stage \
    --model2 models/multi_stage \
    --output results/strategy_comparison.md
  ```

### Phase 7: Distributed Training Validation

Validate that distributed training produces equivalent results with speedup.

- [ ] **Implement distributed validator** (`src/realign/validation/distributed_validator.py`)
  ```python
  class DistributedValidator:
      def validate_distributed_training(
          self,
          model: str,
          dataset: str,
          single_machine_results: Dict
      ) -> Dict:
          """Validate distributed training matches single-machine"""
          
      def measure_speedup(
          self,
          single_machine_time: float,
          distributed_time: float
      ) -> float:
          """Calculate actual speedup achieved"""
  ```

- [ ] **Test distributed training**
  ```bash
  # Single machine baseline
  time python -m realign.validation.train \
    --content-type REASONING \
    --output models/single_machine

  # Distributed (2x Mac Studios)
  time mlx.launch --backend ring --hosts mac-pro,mac-studio \
    python -m realign.validation.train \
      --content-type REASONING \
      --output models/distributed

  # Compare results and timing
  python -m realign.validation.compare_distributed \
    --single models/single_machine \
    --distributed models/distributed \
    --output results/distributed_validation.md
  ```

- [ ] **Validation criteria**
  - Results within 1% of single-machine training
  - Speedup ≥ 1.3x (30% faster)
  - No loss of quality
  - Reproducible results

### Phase 8: Documentation & Reporting

Generate comprehensive validation reports.

- [ ] **Create report generator** (`src/realign/validation/report_generator.py`)
  ```python
  class ValidationReportGenerator:
      def generate_comprehensive_report(
          self,
          validation_results: Dict,
          output_path: str
      ):
          """Generate full validation report"""
          # Executive summary
          # Per-content-type results
          # Regression analysis
          # Multi-stage comparison
          # Distributed training validation
          # Recommendations
          
      def generate_charts(self, results: Dict, output_dir: str):
          """Generate visualization charts"""
          # Improvement bar charts
          # Regression heat maps
          # Training strategy radar charts
          # Speedup comparison
  ```

- [ ] **Report template** (`results/validation/VALIDATION_REPORT.md`)
  ```markdown
  # ReAlign End-to-End Validation Report
  
  ## Executive Summary
  - ✅ 10/10 content types show improvement
  - ✅ Multi-stage training outperforms single-stage by 12%
  - ✅ Distributed training 37% faster with equivalent results
  - ⚠️ Minor regression in FACT after BELIEF training (within limits)
  
  ## Content Type Results
  ### BELIEF (DPO Training)
  - Baseline: 0.62 alignment score
  - Trained: 0.78 alignment score (+25%)
  - Regression check: PASS (max 3% regression in other types)
  
  [... for all 10 types]
  ```

### Phase 9: CI/CD Integration (Optional)

Automate validation in CI/CD pipeline.

- [ ] **Create GitHub Actions workflow** (`.github/workflows/validation.yml`)
  ```yaml
  name: End-to-End Validation
  
  on:
    push:
      branches: [main, develop]
    schedule:
      - cron: '0 0 * * 0'  # Weekly
  
  jobs:
    validate:
      runs-on: self-hosted  # Need Mac Studio
      steps:
        - uses: actions/checkout@v2
        - name: Run validation suite
          run: python -m realign.validation.run_all
        - name: Generate report
          run: python -m realign.validation.report
        - name: Upload results
          uses: actions/upload-artifact@v2
          with:
            name: validation-results
            path: results/validation/
  ```

- [ ] **Nightly validation runs**
  - Run full validation suite overnight
  - Email report if failures
  - Track improvement over time

## Success Criteria

- [ ] All 10 content types tested end-to-end
- [ ] Measurable improvement for each type
- [ ] Regression detection working
- [ ] Multi-stage training validated
- [ ] Distributed training validated
- [ ] Comprehensive reports generated
- [ ] Reproducible results
- [ ] Documentation complete
- [ ] CI/CD integration (optional)

## Example Commands

```bash
# Run full validation suite
python -m realign.validation.run_all \
  --base-model meta-llama/Llama-3.1-70B \
  --output-dir results/validation/run_001 \
  --distributed

# Quick validation (single content type)
python -m realign.validation.quick_test \
  --content-type REASONING \
  --iterations 100 \
  --output results/quick_test.json

# Compare training methods
python -m realign.validation.method_comparison \
  --content-type SKILL \
  --methods lora,dpo,lora+dpo \
  --output results/method_comparison.md
```

## File Structure

```
src/realign/validation/
├── __init__.py
├── baseline_runner.py
├── training_executor.py
├── post_training_evaluator.py
├── regression_detector.py
├── multistage_validator.py
├── distributed_validator.py
├── report_generator.py
└── run_all.py

data/test_datasets/
├── belief_test/
├── fact_test/
├── reasoning_test/
└── ... (all 10 types)

results/validation/
├── run_001/
│   ├── VALIDATION_REPORT.md
│   ├── baselines.json
│   ├── trained_results.json
│   ├── regression_analysis.json
│   ├── multistage_comparison.json
│   ├── distributed_validation.json
│   └── charts/
└── latest -> run_001
```

## References

- **Guide**: [TRAINING_METHODS_BY_CONTENT_TYPE.md](docs/TRAINING_METHODS_BY_CONTENT_TYPE.md)
- **MLX Ecosystem**: [MLX_ECOSYSTEM_GUIDE.md](docs/MLX_ECOSYSTEM_GUIDE.md)
- Related: #49 (Content Type Classification)
- Related: #50 (Benchmark Testing)
- Related: #48 (mlx-lm-lora Integration)

## Labels

`enhancement` `testing` `validation` `v3.2.0`
